Cleanup CMake code to avoid creation of empty Zephyr libraries

tejlmand opened the pull request:
Fixes: #38403
Fixes #38622

This is a series of commits to followup on #37512 by ensuring empty Zephyr libraries are not created.

With this series of commits, then several occurrences of:
```
No SOURCES given to Zephyr library: <foobar>
    
Excluding target from build.
```

has now been fixed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38471)
<!-- Reviewable:end -->

tejlmand committed:
drivers: create BT_DRIVERS Kconfig entry

Fixes: #38403

Changing Bluetooth drivers from being a menu into a menuconfig.
This aligns the Bluetooth driver configuration with other driver
configurations as well as provides a setting which identifies if
Bluetooth drivers has been enable.

This further helps to avoid empty Zephyr libraries for bluetooth
samples.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
drivers: net: adding NET_DRIVERS menuconfig

Fixes: #38403

Adding NET_DRIVERS menuconfig so that network drivers are grouped
together in its own menu entry under drivers, similar to most other
drivers.

This further has the advantages that `CONFIG_NET_DRIVERS` can be used
for testing to determine if network drivers has been selected.

This changed revealed a dependency loop where both `select` (for SLIP)
and `depends` (for PPP) which both depends on NET_DRIVERS` where in use
in the dependency tree for Qemu networking, especially NET_SLIP_TAP.

This is handled by defaulting `NET_DRIVERS` to `y` when building for a
Qemu target.
`SLIP` had a dependency to `!QEMU_TARGET || NET_QEMU_SLIP`. This is
changed so that SLIP prompt depends on `!QEMU_TARGET` which provides
full user control in hardware but makes the symbol promptless on Qemu
targets.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
modules: hal_nxp: removing always empty zephyr_library

No sources were ever added to the `zephyr_library()` defined in
modules/hal_nxp/usb/CMakeLists.txt, thus removing this lib to avoid
the warning:

> No SOURCES given to Zephyr library: modules__hal_nxp__usb
>
> Excluding target from build.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
drivers: ethernet: remove dedicated drivers__ethernet__native_posix lib

Fixes: #38403

The two eth_native_posix.c and eth_native_posix_adapt.c are now added
to the common drivers__ethernet Zephyr library.

Instead of creating a dedicated library for just two files those files
are now added to the common ethernet library, see also #8826.
Instead, the dedicated compile definitions required for those files are
specified using COMPILE_DEFINITIONS on the source files.

This also avoids the following warning as the ethernet library is no
longer empty.

> No SOURCES given to Zephyr library: drivers__ethernet
>
> Excluding target from build.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

dcpleung commented:
Hm... if you revert the commit in `subsys/bluetooth/audio`, it would put those file in `libzephyr.a`. Or am I missing something?

tejlmand commented:
> Hm... if you revert the commit in `subsys/bluetooth/audio`, it would put those file in `libzephyr.a`. Or am I missing something?

ahh, right, because there is no `zephyr_library()` in `subsys/bluetooth`.
I had messed up the location of `subsys/bluetooth/audio` to somehow think it was under `drivers/bluetooth/audio`.

The `drivers/bluetooth/CMakeLists.txt` has a `zephyr_library()` in which case subdirs inherit the library if they do `zephyr_library_sources()`.

I better pay better attention to `subsys` != `drivers`.
Thanks for spotting.

mmahadevan108 commented on drivers/CMakeLists.txt:
```diff
@@ -50,7 +50,7 @@ add_subdirectory_ifdef(CONFIG_PM_CPU_OPS pm_cpu_ops)
 add_subdirectory_ifdef(CONFIG_SERIAL_HAS_DRIVER serial)
 add_subdirectory_ifdef(CONFIG_BT_DRIVERS       bluetooth)
-add_subdirectory_ifdef(CONFIG_NETWORKING       net)
+add_subdirectory_ifdef(CONFIG_NET_DRIVERS      net)
```
Should you select `NET_DRIVERS` inside `subsys/net/Kconfig`

tejlmand replied:
Where exactly would you do that ?

If you take a sample like `bluetooth/ipsp` it will enable networking in its `prj.conf`:
https://github.com/zephyrproject-rtos/zephyr/blob/c8c3d32a0f126e449924fc3cbe0ed5d7d3cbadfa/samples/bluetooth/ipsp/prj.conf#L8

but it will not select any net drivers (loopback/slip/ppp) to enter `zephyr_library()` created here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/net/CMakeLists.txt

Resulting in:
```
  No SOURCES given to Zephyr library: drivers__net

  Excluding target from build.
```

That's why `CONFIG_NET_DRIVERS=y` must explicitly be selected when needed.

tejlmand committed:
drivers: gpio: remove unused GPIO selection

Fixes: #38403

Removing unneeded `imply GPIO` and `CONFIG_GPIO=y` occurrences where no
files are added to the gpio zephyr library.

Also removed `CONFIG_GPIO=y` occurences where this is handled by
defconfigs for the soc or board.

Selection of GPIO without selecting any drivers results in the warning:

> No SOURCES given to Zephyr library: drivers__gpio
>
> Excluding target from build.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
drivers: console: remove unused CONSOLE selection

Fixes: #38403

Removing unneeded `CONFIG_CONSOLES=y` occurrences where no console
driver is selected.

Such selection results in an empty drivers__console zephyr library,
which again results in the following warning message.

> No SOURCES given to Zephyr library: drivers__console
>
> Excluding target from build.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
@dcpleung has removed the revert.
Still looking into how exactly we handle the bt_audio in the cleanest way, because `BT_AUDIO_UNICAST` may be selected as that will configure the stack in specific way, but without adding any files to the lib.

dcpleung commented:
I am wondering if we should simply have one library for the whole bluetooth subsys, instead of a couple smaller ones.

Or, it can be worked around like your suggestion in another issue, where a list of source files is created. And if this list if empty, skip all the `zephyr_library_*()` calls.

tejlmand commented:
> I am wondering if we should simply have one library for the whole bluetooth subsys, instead of a couple smaller ones.

yes, but I would keep that outside of this PR.



> Or, it can be worked around like your suggestion in another issue, where a list of source files is created. And if this list if empty, skip all the `zephyr_library_*()` calls.

doing so is simply hiding bad Kconfig / config file settings.
I think it's good if we can clean up situations like the GPIO being enabled without selecting any GPIO driver.

Or the pinmux case where pinmux is enabled board the board CMakeLists.txt places the pinmux.c in a completely  different library.
In such case the pinmux.c ought to be placed in the pinmux driver lib, especially considering that some IDEs are presenting files in their libs. With those IDEs it's annoying if files are "jumping" around just because you build for another board.

(Some IDEs shows the files based on the folder structure on the disk, and are not affected by which lib a source file is end up)

dcpleung commented:
Personally, I prefer to have one bluetooth subsys library. Though it would be up to the bluetooth maintainers on what they want to do.

tejlmand commented:
> Personally, I prefer to have one bluetooth subsys library. Though it would be up to the bluetooth maintainers on what they want to do.

Or just a few. It could also be argued that having a bluetooth controller library + bluetooth host library would make sense.
As long as it has been consciously chosen, and not because `zephyr_library()` is just everywhere else then more than one library can be justified. But as you also mention, it's for the maintainers of the subsystem to decide, we can just guide them :wink:

dleach02 commented:
> Or just a few. It could also be argued that having a bluetooth controller library + bluetooth host library would make sense.
> As long as it has been consciously chosen, and not because `zephyr_library()` is just everywhere else then more than one library can be justified. But as you also mention, it's for the maintainers of the subsystem to decide, we can just guide them 😉

Separating controller from host makes a lot of sense. +1

rlubos commented on subsys/net/ip/Kconfig:
```diff
@@ -589,10 +589,10 @@ config NET_SLIP_TAP
 	bool "TAP SLIP driver"
 	depends on NET_QEMU_SLIP
 	depends on NET_NATIVE
-	select SLIP
```
This breaks the default configuration for practically any networking sample that can run on qemu, it can't go as it is. If we want to give up on `select`, it's needed to update the default value of `SLIP` config, likely set to yes if `NET_QEMU_SLIP` is enabled.

rlubos replied:
Same for the `NET_DRIVERS`, it seems it should also be enabled by default for qemu networking.

tejlmand replied:
@rlubos  interesting, seems I missed something.
How come networking samples are not failing with twister when `SLIP` is disabled, even when running the integration tests ?

Have been running complete CI for boards, and integration tests for `qemu_x86`.

A re-run just for `tests/net`.
```
$ ./scripts/twister -c -N --inline-logs -T tests/net/ -p qemu_x86 -G
ZEPHYR_BASE unset, using "/projects/github/ncs/zephyr"
Deleting output directory /projects/github/ncs/zephyr/twister-out
INFO    - Zephyr version: v2.7.0-rc1-140-g76dc66548970
INFO    - JOBS: 16
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testcase list...
INFO    - 152 test scenarios (152 configurations) selected, 59 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    - Total complete:   93/  93  100%  skipped:   63, failed:    0
INFO    - 89 of 89 test configurations passed (100.00%), 0 failed, 63 skipped with 0 warnings in 94.45 seconds
INFO    - In total 725 test cases were executed, 1653 skipped on 1 out of total 394 platforms (0.25%)
INFO    - 81 test configurations executed on platforms, 8 test configurations were only built.
INFO    - Saving reports...
```

I would expect tests that require `CONFIG_SLIP=y` to fail in such case.
But all `tests/net` passes both on main and with this PR, even test cases where `CONFIG_SLIP=y` in main but not in PR.

rlubos replied:
As far as I understand, the networking tests executed by twister are self-contained, i.e. using the loopback interface, so they don't need the actual driver in order to execute. But when it comes to the samples, where the actual network communication is expected, they won't work w/o SLIP (by default). 

The tricky part is, that even w/o SLIP being enabled the samples build just fine, there's no compilation error. But when you run the sample, the system complains that there is no network interface to work with.

tejlmand replied:
> But when it comes to the samples, where the actual network communication is expected, they won't work w/o SLIP (by default).

yes, but this line in the setting that was changed from `select` to `depends` looks like this:
```
config NET_SLIP_TAP
	bool "TAP SLIP driver"
	depends on NET_QEMU_SLIP
	depends on NET_NATIVE
	depends on SLIP
```
so it has a dependency to `NET_QEMU_SLIP`, which is defined here: 
https://github.com/zephyrproject-rtos/zephyr/blob/00f5f3fab4c48097b1c2bc49698137fc50bcc90d/subsys/net/ip/Kconfig#L74

and is under a choice which `depends on QEMU_TARGET`:
https://github.com/zephyrproject-rtos/zephyr/blob/00f5f3fab4c48097b1c2bc49698137fc50bcc90d/subsys/net/ip/Kconfig#L64-L74

so my change is only applicable for Qemu.

And twister launches Qemu in order to run those tests:
```
2021-09-22 17:03:37,246 - twister - DEBUG - Running qemu_x86/tests/net/neighbor/net.neighbor.preempt (qemu)
2021-09-22 17:03:37,247 - twister - DEBUG - QEMU (637692): Booting from ROM..*** Booting Zephyr OS build v2.7.0-rc1-140-g76dc66548970  ***
2021-09-22 17:03:37,248 - twister - DEBUG - QEMU (637692): Running test suite test_arp_fn
2021-09-22 17:03:37,248 - twister - DEBUG - QEMU (637692): ===================================================================
2021-09-22 17:03:37,248 - twister - DEBUG - QEMU (637692): START - test_arp
2021-09-22 17:03:37,248 - twister - DEBUG - Spawning QEMUHandler Thread for qemu_x86/tests/net/neighbor/net.neighbor.preempt
2021-09-22 17:03:37,249 - twister - DEBUG - Running qemu_x86/tests/net/ieee802154/fragment/net.ieee802154.fragment (qemu)
2021-09-22 17:03:37,250 - twister - DEBUG - Spawning QEMUHandler Thread for qemu_x86/tests/net/ieee802154/fragment/net.ieee802154.fragment
2021-09-22 17:03:37,251 - twister - DEBUG - Running qemu_x86/tests/net/igmp/net.igmp.preempt (qemu)
2021-09-22 17:03:37,254 - twister - DEBUG - QEMU (637692): PASS - test_arp in 0.7 seconds
2021-09-22 17:03:37,255 - twister - DEBUG - QEMU (637692): ===================================================================
2021-09-22 17:03:37,255 - twister - DEBUG - QEMU (637692): Test suite test_arp_fn succeeded
2021-09-22 17:03:37,255 - twister - DEBUG - QEMU (637692): ===================================================================
2021-09-22 17:03:37,255 - twister - DEBUG - QEMU (637692): PROJECT EXECUTION SUCCESSFUL
....
```

so if qemu anyway just uses the loopback device, then why is the `NET_QEMU_SLIP` flag there at all ?

>  But when it comes to the samples, where the actual network communication is expected, they won't work w/o SLIP (by default).

Are you having some other samples in mind besides `tests/net` ?
Notice, the `select SLIP` we are discussing here is only applicable to the `QEMU_TARGET` due to dependency chain.
This change has no impact on any real board.

> The tricky part is, that even w/o SLIP being enabled the samples build just fine, there's no compilation error. But when you run the sample, the system complains that there is no network interface to work with.

yes, which is exactly why i'm running the samples using qemu.

rlubos replied:
> Are you having some other samples in mind besides tests/net ?

Every sample under `samples/net`.

> so if qemu anyway just uses the loopback device, then why is the NET_QEMU_SLIP flag there at all ?

It uses a loopback interface for tests. But if you run some actual sample, that needs to communicate with some other device in the network, you need some emulated driver (and by default SLIP was used).

> Notice, the select SLIP we are discussing here is only applicable to the QEMU_TARGET due to dependency chain.
This change has no impact on any real board.

And this is perfectly fine, real boards should use some actual networking hardware they are providing. But having a possibility to run networking samples in emulated environment like qemu easily (i.e. with the default `prj.conf`, w/o any extra hassle) is a must-have feature IMO, dropping such support is a serious regression.

tejlmand replied:
@rlubos thanks for the detailed explanation.

I have pushed an update that changes some Kconfig slightly, but in a way so that things should work as desired now.

rlubos replied:
Thanks, it seems to work fine now

brockus-zephyr reviewed (approved):
> Reviewed 2 of 3 files at r1, 81 of 81 files at r2, 17 of 17 files at r3, 67 of 72 files at r4, 9 of 9 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38471)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @albertofloyd, @andyross, @GTLin08, @IRISZZW, @katsuster, @keith-zephyr, @mengxianglinx, @mgielda, @mnkp, @nashif, @scottwcpg, @sjg20, and @VenkatKotakonda)



<!-- Sent from Reviewable.io -->

rlubos reviewed (approved):
Thanks, networking samples work as they should now.
